make torch.testing asserts importable

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55385 Add support for checking tensor containers in `torch.testing`
* #55145 make torch.testing error messages more expressive
* #54784 add tests for torch.testing
* #54781 torch.testing allclose -> close
* #54780 change torch.testing helper asserts to checks
* **#54769 make torch.testing asserts importable**

Follow-up to #53820. This

- makes the `asserts.py` module private as per suggestion from @rgommers in https://github.com/pytorch/pytorch/pull/53820#issuecomment-802661387. With this the functions should only be accessible through `torch.testing`, giving us the option the change the underlying structure later.
- moves the code from `torch/testing/__init__.py` to `torch/testing/_core.py` (happy to accept other name suggestions). Otherwise we can't import the new `_asserts.py` in `torch/testing/__init__.py` due to circular imports.
- factors out some functionality from `torch/testing/_internal/common_utils.py` that is needed in new `_asserts.py` to `torch/testing/_internal/testing_utils.py` to not be dependent on `numpy` at import.

Differential Revision: [D27438451](https://our.internmc.facebook.com/intern/diff/D27438451)

pmeier committed:
make torch.testing asserts importable

move coda out of torch.testing.__init__
make everything importable again


delete torch/testing/asserts.py


please flake8


provide more stuff in torch.testing


provide even more stuff in torch.testing


expose more


fix import


make private


expose more


fix mypy

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing asserts importable"


move coda out of torch.testing.__init__
make everything importable again


delete torch/testing/asserts.py


please flake8


provide more stuff in torch.testing


provide even more stuff in torch.testing


expose more


fix import


make private


expose more


fix mypy

[ghstack-poisoned]

pmeier commented:
@mruberry Not sure about the test failures. For some reason importing `numpy` is failing. From the error message 

```
ModuleNotFoundError: Sorry PyTorch, but our NumPy is in the other folder
```

it seems someone planned for this. Do you have insights?

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -42,10 +42,9 @@
-from torch.testing import \
-    (_compare_tensors_internal, _compare_scalars_internal, _compare_return_type,
-     floating_types_and, integral_types, complex_types)
+from .._core import _compare_tensors_internal, _compare_scalars_internal, _compare_return_type
```
Style nit: keep the parentheses; formatting becomes a pain without them

mruberry commented:
> @mruberry Not sure about the test failures. For some reason importing `numpy` is failing. From the error message
> 
> ```
> ModuleNotFoundError: Sorry PyTorch, but our NumPy is in the other folder
> ```
> 
> it seems someone planned for this. Do you have insights?

I'd spend some more time debugging (possibly after a rebase for good measure). It looks like two tests are failing. The ModuleNotFoundError is coming from a "fake NumPy" that was added for testing purposes in February. What's causing it to fail?

pmeier commented:
#52794 added a test to check that `torch` is importable 

> if PyTorch was compiled with wrong NumPy version, of NumPy is not commonly available on the platform (which is often the case on AARCH64 or Apple M1)

The new `_asserts.py` needs functionality from `torch.testing._internal` which imports `numpy` unconditionally. Since this has not happened before, the tests went fine. I'll factor out the functionality that we need into a separate file.

pmeier committed:
Update on "make torch.testing asserts importable"


Follow-up to #53820. This

- makes the `asserts.py` module private as per suggestion from @rgommers in https://github.com/pytorch/pytorch/pull/53820#issuecomment-802661387. With this the functions should only be accessible through `torch.testing`, giving us the option the change the underlying structure later.
- moves the code from `torch/testing/__init__.py` to `torch/testing/_core.py` (happy to accept other name suggestions). Otherwise we can't import the new `_asserts.py` in `torch/testing/__init__.py` due to circular imports.

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing asserts importable"


Follow-up to #53820. This

- makes the `asserts.py` module private as per suggestion from @rgommers in https://github.com/pytorch/pytorch/pull/53820#issuecomment-802661387. With this the functions should only be accessible through `torch.testing`, giving us the option the change the underlying structure later.
- moves the code from `torch/testing/__init__.py` to `torch/testing/_core.py` (happy to accept other name suggestions). Otherwise we can't import the new `_asserts.py` in `torch/testing/__init__.py` due to circular imports.

[ghstack-poisoned]

mruberry reviewed (approved):
Cool; so this also fixes that importing torch imports numpy?

I wonder if we should add a test for that?

mruberry commented:
> #52794 added a test to check that `torch` is importable
> 
> > if PyTorch was compiled with wrong NumPy version, of NumPy is not commonly available on the platform (which is often the case on AARCH64 or Apple M1)
> 
> The new `_asserts.py` needs functionality from `torch.testing._internal` which imports `numpy` unconditionally. Since this has not happened before, the tests went fine. I'll factor out the functionality that we need into a separate file.

I like the comment you made in the code about this. Would you update this PR description, too? We should talk about our plan for the structure of all this code at our next testing sync next week.

pmeier commented:
> Cool; so this also fixes that importing torch imports numpy?

Correct.

> I wonder if we should add a test for that?

The test that failed before tests for exactly that:

https://github.com/pytorch/pytorch/blob/444e5f0b60358c3ff17204938eb17e4a3b31ad4c/.jenkins/pytorch/test.sh#L160-L164

Still, the error message through me off

https://github.com/pytorch/pytorch/blob/444e5f0b60358c3ff17204938eb17e4a3b31ad4c/.jenkins/pytorch/fake_numpy/numpy.py#L1

Maybe we can change it into something more descriptive like

```python
raise ModuleNotFoundError("Importing 'torch' must not require 'numpy' to be present.")
```

Cc @malfet

mruberry commented:
I need to investigate this further internally. The new structure either requires some internal tweaks or there's a problem with the specifying the imports relative to the current location.

Example failure: 

```
  File ".../tree/torch/testing/_asserts.py", line 7, in <module>
    from ._internal.testing_utils import (
ModuleNotFoundError: No module named 'torch.testing._internal'
```

pmeier committed:
Update on "make torch.testing asserts importable"


Follow-up to #53820. This

- makes the `asserts.py` module private as per suggestion from @rgommers in https://github.com/pytorch/pytorch/pull/53820#issuecomment-802661387. With this the functions should only be accessible through `torch.testing`, giving us the option the change the underlying structure later.
- moves the code from `torch/testing/__init__.py` to `torch/testing/_core.py` (happy to accept other name suggestions). Otherwise we can't import the new `_asserts.py` in `torch/testing/__init__.py` due to circular imports.
- factors out some functionality from `torch/testing/_internal/common_utils.py` that is needed in new `_asserts.py` to `torch/testing/_internal/testing_utils.py` to not be dependent on `numpy` at import.

Differential Revision: [D27438451](https://our.internmc.facebook.com/intern/diff/D27438451)

[ghstack-poisoned]

mruberry commented:
I had some more time to investigate the internal build system, and I think the issue is simply that the current build system isn't designed to allow _internal to be imported into testing, since torch.testing is always imported but torch.testing._internal is only imported on test builds. 

We could try and tweak the internal build system, but I'd prefer we just work around the issue. Can we just copy parts of _internal we need for asserts.py into torch/testing/asserts.py for now, for example?

pmeier committed:
Update on "make torch.testing asserts importable"


Follow-up to #53820. This

- makes the `asserts.py` module private as per suggestion from @rgommers in https://github.com/pytorch/pytorch/pull/53820#issuecomment-802661387. With this the functions should only be accessible through `torch.testing`, giving us the option the change the underlying structure later.
- moves the code from `torch/testing/__init__.py` to `torch/testing/_core.py` (happy to accept other name suggestions). Otherwise we can't import the new `_asserts.py` in `torch/testing/__init__.py` due to circular imports.
- factors out some functionality from `torch/testing/_internal/common_utils.py` that is needed in new `_asserts.py` to `torch/testing/_internal/testing_utils.py` to not be dependent on `numpy` at import.

Differential Revision: [D27438451](https://our.internmc.facebook.com/intern/diff/D27438451)

[ghstack-poisoned]

pmeier commented:
> Can we just copy parts of _internal we need for asserts.py into torch/testing/asserts.py for now, for example?

Yes, we can. I've removed the `torch/testing/_internal/testing_utils.py` again and simply copied the dtype precisions. Note that copying `get_comparison_dtype` was not necessary, since the todos are outdated.

https://github.com/pytorch/pytorch/blob/df299dbd7d8241669aaac7ce07ed0c034f219b4f/torch/testing/_internal/common_utils.py#L858-L873

Thus, it can simply replaced by `torch.promote_types()`.

mruberry commented:
Let's see what the internal build system thinks.
